packages/process/process_monitor.pony

```diff
@@ -295,20 +305,37 @@ actor ProcessMonitor
   fun _child(
     path: String,
     argp: Array[Pointer[U8] tag],
-    envp: Array[Pointer[U8] tag])
+    envp: Array[Pointer[U8] tag],
+    wdir: (FilePath | None))
   =>
     """
-    We are now in the child process. We redirect STDIN, STDOUT and STDERR
-    to their pipes and execute the command. The command is executed via
-    execve which does not return on success, and the text, data, bss, and
-    stack of the calling process are overwritten by that of the program
-    loaded. We've set the FD_CLOEXEC flag on all file descriptors to ensure
-    that they are all closed automatically once @execve gets called.
+    We are now in the child process. If a working directory was passed to
+    the child, a chdir is attempted. We redirect STDIN, STDOUT and STDERR to
+    their pipes and execute the command. The command is executed via execve which
+    does not return on success, and the text, data, bss, and stack of the
+    calling process are overwritten by that of the program loaded. We've set the
+    FD_CLOEXEC flag on all file descriptors to ensure that they are all closed
+    automatically once @execve gets called.
     """
     ifdef posix then
+      match wdir
+      | let d: FilePath =>
+        let dir: Pointer[U8] tag = d.path.cstring()
+
+        if 0 > @chdir[I32](dir) then
+          let errno: I32 = @pony_os_errno()
+          let strerror = @strerror[Pointer[U8]](errno)
+          let size = @strlen[USize](strerror)
+          let len = @write[ISize](_errpipe_write, strerror, size)
+          @_exit[None](I32(71)) // EX_OSERR from sysexits.h
+        end
+      | None => None
+      end
       _dup2(_stdin_read, _STDINFILENO())    // redirect stdin
       _dup2(_stdout_write, _STDOUTFILENO()) // redirect stdout
       _dup2(_stderr_write, _STDERRFILENO()) // redirect stderr
+
+      @close[I32](_errpipe_write)
       if 0 > @execve[I32](path.cstring(), argp.cpointer(),
         envp.cpointer())
       then
```

Comment:
I think we could also signal an error when `execve` fails to execute the given executable and send the `errno` through the `err_pipe`. It might be necessary to somehow mark, what kind of error happened by sending another number or something. Because we need to know from the reading side, what kind of error to report to `ProcessNotify.failed`. If the `errno`s are possibly overlapping, we might need to differentiate, maybe.

We could issue an `ExecveError` to `ProcessNotify.failed` here.